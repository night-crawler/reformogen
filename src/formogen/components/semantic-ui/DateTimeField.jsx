import React from 'react';

import moment from 'moment';

import _ from 'lodash';

import { Form } from 'semantic-ui-react';

import DatePicker from 'react-datepicker';

import { MessageList } from './MiscComponents';
import Label from './Label';
import propTypes from '../fieldPropTypes';


/**
 * Component should receive a value in ISO format
 * @param {object} props
 * @param {string} props.name - internal field name
 * @param {string} props.verbose_name
 * @param {string} props.help_text -
 * @param {bool} props.blank - determines if we can leave field unfilled
 * @param {bool} props.null - determines if we can store NULL values
 * @param {bool} props.editable - can we edit the field value?
 * @param {string} props.type - Django ORM field type
 * @param {bool} props.required
 * @param {string} props.value - well ISO-formatted value to pass moment()
 * @returns {XML}
 * @constructor
 */
DateTimeField.propTypes = propTypes;
export default function DateTimeField(props) {
    const handleChange = (momentTimeObject, e) => {
        // proxy event with SUI-React compliant style
        props.onChange(e, {
            name: props.name,
            value: momentTimeObject ? momentTimeObject.toISOString() : null
        });
    };

    let _props = {
        name: props.name,
        selected: props.value ? moment(props.value) : null,  // null == no selected value
        onChange: handleChange,
        showTimeSelect: true,
        timeFormat: 'HH:mm',
        timeIntervals: 60,
        locale: props.locale,
        placeholderText: props.placeholder,
        todayButton: 'Now',
        dateFormat: 'YYYY.MM.DD HH:mm',
        showYearDropdown: true,
        showMonthDropdown: true,
        dropdownMode: 'select',
    };

    if (_.isFunction(props.updateProps)) {
        _props = props.updateProps(_props, props);
    }

    return (
        <Form.Field
            required={ props.required }
            disabled={ !props.editable }
            width={ props.layoutOpts.width }
            error={ !_.isEmpty(props.errors) }
        >
            <Label { ...props } />
            <DatePicker { ..._props } />
            { !props.helpTextOnHover ? <span className="help-text">{ props.help_text }</span> : ''}
            <MessageList messages={ props.errors } />
        </Form.Field>
    );
}
