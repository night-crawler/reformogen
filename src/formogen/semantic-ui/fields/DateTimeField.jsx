import React from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';
import moment from 'moment';
import { Form } from 'semantic-ui-react';
import DatePicker from 'react-datepicker';

import { errorsType, layoutOptsType } from '../../fieldPropTypes';
import { FieldLabel } from '../common/FieldLabel';
import { ErrorsList } from '../common/ErrorsList';


DateTimeField.propTypes = {
  type: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.instanceOf(Date)]),
  max_length: PropTypes.number,
  help_text: PropTypes.string.isRequired,
  placeholder: PropTypes.string,
  errors: errorsType,

  required: PropTypes.bool,
  editable: PropTypes.bool,

  helpTextOnHover: PropTypes.bool,
  layoutOpts: layoutOptsType,
  locale: PropTypes.string,

  
  timeFormat: PropTypes.string,
  timeIntervals: PropTypes.number,
  todayButton: PropTypes.string,
  dateFormat: PropTypes.string,
  dropdownMode: PropTypes.string,

  onChange: PropTypes.func,
};
DateTimeField.defaultProps = {
  timeFormat: 'HH:mm',
  timeIntervals: 30,
  todayButton: 'Now',
  dateFormat: 'YYYY.MM.DD HH:mm',
  dropdownMode: 'select',
};
export function DateTimeField(props) {
  const handleChange = (momentTimeObject, e) => {
    props.onChange(e, {
      name: props.name,
      value: momentTimeObject ? momentTimeObject.toISOString() : null
    });
  };

  return (
    <Form.Field
      required={ props.required }
      disabled={ !props.editable }
      width={ props.layoutOpts.width }
      error={ !_.isEmpty(props.errors) }
    >
      <FieldLabel { ...props } />
      <DatePicker 
        name={ props.name }
        selected={ props.value ? moment(props.value) : null }  // null == no selected value
        onChange={ handleChange }
        showTimeSelect={ true }
        
        locale={ props.locale }
        placeholderText={ props.placeholder }
        timeFormat={ props.timeFormat }
        timeIntervals={ props.timeIntervals }
        todayButton={ props.todayButton }
        // TODO: get from moment
        dateFormat={ props.dateFormat }
        dropdownMode={ props.dropdownMode }
        
        showYearDropdown={ true }
        showMonthDropdown={ true }
      />
      { !props.helpTextOnHover
        ? <span className='help-text'>{ props.help_text }</span>
        : ''
      }
      <ErrorsList messages={ props.errors } />
    </Form.Field>
  );
}
