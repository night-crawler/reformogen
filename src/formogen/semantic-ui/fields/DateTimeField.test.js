import { DateTimeField } from './DateTimeField';


it('<DateTimeField />', () => {
  const wrapper = shallow(
    <DateTimeField
      type='DateTimeField'
      name='DateTimeField: name'
      verbose_name='DateTimeField: verbose_name'
      help_text='DateTimeField: help_text'
      layoutOpts={ { width: 1 } }
    />
  );
  expect(wrapper).toMatchSnapshot();
});
