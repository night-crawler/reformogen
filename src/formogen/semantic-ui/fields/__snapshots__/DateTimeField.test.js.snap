// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DateTimeField /> 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <DateTimeField
    dateFormat="YYYY.MM.DD HH:mm"
    dropdownMode="select"
    help_text="DateTimeField: help_text"
    layoutOpts={
      Object {
        "width": 1,
      }
    }
    name="DateTimeField: name"
    timeFormat="HH:mm"
    timeIntervals={30}
    todayButton="Now"
    type="DateTimeField"
    verbose_name="DateTimeField: verbose_name"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <FieldLabel
          dateFormat="YYYY.MM.DD HH:mm"
          dropdownMode="select"
          help_text="DateTimeField: help_text"
          layoutOpts={
            Object {
              "width": 1,
            }
          }
          name="DateTimeField: name"
          timeFormat="HH:mm"
          timeIntervals={30}
          todayButton="Now"
          type="DateTimeField"
          verbose_name="DateTimeField: verbose_name"
        />,
        <DatePicker
          allowSameDay={false}
          dateFormat="YYYY.MM.DD HH:mm"
          dateFormatCalendar="MMMM YYYY"
          disabled={false}
          disabledKeyboardNavigation={false}
          dropdownMode="select"
          monthsShown={1}
          name="DateTimeField: name"
          nextMonthButtonLabel="Next month"
          onBlur={[Function]}
          onChange={[Function]}
          onClickOutside={[Function]}
          onFocus={[Function]}
          onInputClick={[Function]}
          onKeyDown={[Function]}
          onMonthChange={[Function]}
          onSelect={[Function]}
          onYearChange={[Function]}
          preventOpenOnFocus={false}
          previousMonthButtonLabel="Previous Month"
          readOnly={false}
          selected={null}
          shouldCloseOnSelect={true}
          showMonthDropdown={true}
          showTimeSelect={true}
          showYearDropdown={true}
          timeCaption="Time"
          timeFormat="HH:mm"
          timeIntervals={30}
          todayButton="Now"
          withPortal={false}
        />,
        <span
          className="help-text"
        >
          DateTimeField: help_text
        </span>,
        <ErrorsList
          color="red"
          header="Errors"
        />,
      ],
      "disabled": true,
      "error": false,
      "required": undefined,
      "width": 1,
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "dateFormat": "YYYY.MM.DD HH:mm",
          "dropdownMode": "select",
          "help_text": "DateTimeField: help_text",
          "layoutOpts": Object {
            "width": 1,
          },
          "name": "DateTimeField: name",
          "timeFormat": "HH:mm",
          "timeIntervals": 30,
          "todayButton": "Now",
          "type": "DateTimeField",
          "verbose_name": "DateTimeField: verbose_name",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "allowSameDay": false,
          "dateFormat": "YYYY.MM.DD HH:mm",
          "dateFormatCalendar": "MMMM YYYY",
          "disabled": false,
          "disabledKeyboardNavigation": false,
          "dropdownMode": "select",
          "locale": undefined,
          "monthsShown": 1,
          "name": "DateTimeField: name",
          "nextMonthButtonLabel": "Next month",
          "onBlur": [Function],
          "onChange": [Function],
          "onClickOutside": [Function],
          "onFocus": [Function],
          "onInputClick": [Function],
          "onKeyDown": [Function],
          "onMonthChange": [Function],
          "onSelect": [Function],
          "onYearChange": [Function],
          "placeholderText": undefined,
          "preventOpenOnFocus": false,
          "previousMonthButtonLabel": "Previous Month",
          "readOnly": false,
          "selected": null,
          "shouldCloseOnSelect": true,
          "showMonthDropdown": true,
          "showTimeSelect": true,
          "showYearDropdown": true,
          "timeCaption": "Time",
          "timeFormat": "HH:mm",
          "timeIntervals": 30,
          "todayButton": "Now",
          "withPortal": false,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "DateTimeField: help_text",
          "className": "help-text",
        },
        "ref": null,
        "rendered": "DateTimeField: help_text",
        "type": "span",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "color": "red",
          "header": "Errors",
          "messages": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <FieldLabel
            dateFormat="YYYY.MM.DD HH:mm"
            dropdownMode="select"
            help_text="DateTimeField: help_text"
            layoutOpts={
              Object {
                "width": 1,
              }
            }
            name="DateTimeField: name"
            timeFormat="HH:mm"
            timeIntervals={30}
            todayButton="Now"
            type="DateTimeField"
            verbose_name="DateTimeField: verbose_name"
          />,
          <DatePicker
            allowSameDay={false}
            dateFormat="YYYY.MM.DD HH:mm"
            dateFormatCalendar="MMMM YYYY"
            disabled={false}
            disabledKeyboardNavigation={false}
            dropdownMode="select"
            monthsShown={1}
            name="DateTimeField: name"
            nextMonthButtonLabel="Next month"
            onBlur={[Function]}
            onChange={[Function]}
            onClickOutside={[Function]}
            onFocus={[Function]}
            onInputClick={[Function]}
            onKeyDown={[Function]}
            onMonthChange={[Function]}
            onSelect={[Function]}
            onYearChange={[Function]}
            preventOpenOnFocus={false}
            previousMonthButtonLabel="Previous Month"
            readOnly={false}
            selected={null}
            shouldCloseOnSelect={true}
            showMonthDropdown={true}
            showTimeSelect={true}
            showYearDropdown={true}
            timeCaption="Time"
            timeFormat="HH:mm"
            timeIntervals={30}
            todayButton="Now"
            withPortal={false}
          />,
          <span
            className="help-text"
          >
            DateTimeField: help_text
          </span>,
          <ErrorsList
            color="red"
            header="Errors"
          />,
        ],
        "disabled": true,
        "error": false,
        "required": undefined,
        "width": 1,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "dateFormat": "YYYY.MM.DD HH:mm",
            "dropdownMode": "select",
            "help_text": "DateTimeField: help_text",
            "layoutOpts": Object {
              "width": 1,
            },
            "name": "DateTimeField: name",
            "timeFormat": "HH:mm",
            "timeIntervals": 30,
            "todayButton": "Now",
            "type": "DateTimeField",
            "verbose_name": "DateTimeField: verbose_name",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "allowSameDay": false,
            "dateFormat": "YYYY.MM.DD HH:mm",
            "dateFormatCalendar": "MMMM YYYY",
            "disabled": false,
            "disabledKeyboardNavigation": false,
            "dropdownMode": "select",
            "locale": undefined,
            "monthsShown": 1,
            "name": "DateTimeField: name",
            "nextMonthButtonLabel": "Next month",
            "onBlur": [Function],
            "onChange": [Function],
            "onClickOutside": [Function],
            "onFocus": [Function],
            "onInputClick": [Function],
            "onKeyDown": [Function],
            "onMonthChange": [Function],
            "onSelect": [Function],
            "onYearChange": [Function],
            "placeholderText": undefined,
            "preventOpenOnFocus": false,
            "previousMonthButtonLabel": "Previous Month",
            "readOnly": false,
            "selected": null,
            "shouldCloseOnSelect": true,
            "showMonthDropdown": true,
            "showTimeSelect": true,
            "showYearDropdown": true,
            "timeCaption": "Time",
            "timeFormat": "HH:mm",
            "timeIntervals": 30,
            "todayButton": "Now",
            "withPortal": false,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "DateTimeField: help_text",
            "className": "help-text",
          },
          "ref": null,
          "rendered": "DateTimeField: help_text",
          "type": "span",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "color": "red",
            "header": "Errors",
            "messages": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
